libminifi/test/unit/SchedulingAgentTests.cpp

```diff
@@ -36,133 +38,167 @@ class CountOnTriggersProcessor : public minifi::core::Processor {
+  cron_driven_agent->start();
+  auto first_task_reschedule_info = cron_driven_agent->run(count_proc_.get(), context_, factory_);
+  CHECK(!first_task_reschedule_info.isFinished());
+  CHECK(first_task_reschedule_info.getNextExecutionTime() <= std::chrono::steady_clock::now() + 1s);
+  CHECK(count_proc_->getNumberOfTriggers() == 0);
+
+  std::this_thread::sleep_until(first_task_reschedule_info.getNextExecutionTime());
+  auto second_task_reschedule_info = cron_driven_agent->run(count_proc_.get(), context_, factory_);
+  CHECK(!second_task_reschedule_info.isFinished());
+  CHECK(second_task_reschedule_info.getNextExecutionTime() <= std::chrono::steady_clock::now() + 1s);
+  CHECK(count_proc_->getNumberOfTriggers() == 1);
+}
+
+TEST_CASE_METHOD(SchedulingAgentTestFixture, "Cron Driven no future triggers") {
+  count_proc_->setCronPeriod("* * * * * * 2012");
+  auto cron_driven_agent = std::make_shared<CronDrivenSchedulingAgent>(gsl::make_not_null(controller_services_provider_.get()), test_repo_, test_repo_, content_repo_, configuration_, thread_pool_);
+  cron_driven_agent->start();
+  auto first_task_reschedule_info = cron_driven_agent->run(count_proc_.get(), context_, factory_);
+  CHECK(first_task_reschedule_info.isFinished());
+}
+
+TEST_CASE_METHOD(SchedulingAgentTestFixture, "Timer driven should wait for scheduled run even if it yielded") {
+  count_proc_->setSchedulingPeriod(1min);
+  count_proc_->setYieldPeriodMsec(10ms);
+  count_proc_->setShouldYield(true);
+  auto timer_driven_agent = std::make_shared<TimerDrivenSchedulingAgent>(gsl::make_not_null(controller_services_provider_.get()), test_repo_, test_repo_, content_repo_, configuration_, thread_pool_);
+  timer_driven_agent->start();
+  auto first_task_reschedule_info = timer_driven_agent->run(count_proc_.get(), context_, factory_);
+  CHECK(!first_task_reschedule_info.isFinished());
+  CHECK(first_task_reschedule_info.getNextExecutionTime() > std::chrono::steady_clock::now() + 100ms);
+  CHECK(count_proc_->getNumberOfTriggers() == 1);
+}
+
+TEST_CASE_METHOD(SchedulingAgentTestFixture, "Timer driven should wait for yield if processor yielded") {
+  count_proc_->setSchedulingPeriod(10ms);
+  count_proc_->setYieldPeriodMsec(1min);
+  count_proc_->setShouldYield(true);
+  auto timer_driven_agent = std::make_shared<TimerDrivenSchedulingAgent>(gsl::make_not_null(controller_services_provider_.get()), test_repo_, test_repo_, content_repo_, configuration_, thread_pool_);
+  timer_driven_agent->start();
+  auto first_task_reschedule_info = timer_driven_agent->run(count_proc_.get(), context_, factory_);
+  CHECK(!first_task_reschedule_info.isFinished());
+  CHECK(first_task_reschedule_info.getNextExecutionTime() > std::chrono::steady_clock::now() + 100ms);
+  CHECK(count_proc_->getNumberOfTriggers() == 1);
+}
```

Comment:
I don't insist if you think it's clearer this way, but these two tests are identical except for the first two lines, so they could be merged into a single test with two SECTIONs around those pairs of lines.  To me, that would be more readable.

Reply:
It just didnt occur to me, but it makes perfect sense to merge them, good idea https://github.com/apache/nifi-minifi-cpp/pull/1653/commits/c2d2ea6dbecaec9db5750b3d097878735d90d1eb